src/rust/engine/fs/src/snapshot.rs

```diff
@@ -341,6 +341,66 @@ impl Snapshot {
       .to_boxed()
   }
 
+  pub fn strip_prefix(
+    store: Store,
+    root_digest: Digest,
+    prefix: PathBuf,
+  ) -> impl Future<Item = Digest, Error = String> {
+    let store2 = store.clone();
+    Self::get_directory_or_err(store.clone(), root_digest)
+      .and_then(move |dir| {
+        futures::future::loop_fn(
+          (dir, PathBuf::new(), prefix),
+          move |(dir, already_stripped, prefix)| {
+            let mut components = prefix.components();
+            let component_to_strip = components.next();
+            if let Some(component_to_strip) = component_to_strip {
```

Comment:
This surprises me and it looks like a bug, but I suspect that's due to my lack of Rust knowledge.

Reply:
Meanwhile, this makes total sense to me as a Rust newb:

```rust
if already_stripped.components().next().is_some()
```

Reply:
So, there are a couple of things that are weird but useful here...

In another language without such strong move semantics, you'd probably use different naming here, more like:
```rust
let maybe_component_to_strip = components.next();
if let Some(component_to_strip) = maybe_component_to_strip {
```

where the variable names show that the RHS there is an option, and the new variable `component_to_strip` is not an option; it's destructuring out the concrete value from the option. Because the `if let` consumes the variable on the RHS (that variable no longer exists), we often use the same name for them; I'm not sure whether I like that as a pattern or not...

But it's more concise than:
```rust
let maybe_component_to_strip = components.next();
if maybe_component_to_strip.is_some() {
  let component_to_strip = maybe_component_to_strip.unwrap();
}
```

Significantly, it's also more _correct_ - any time you have a `.unwrap()` you need to convince yourself both that it's safe (the value is definitely present), and that any future change to surrounding context preserves that safety (you could trivially violate that by changing the `is_some` to `is_none`); using the destructuring form removes this; you have exactly one check, and it can't get inconsistent with itself.

So in summary: I strongly prefer the destructuring over the `is_some()` check, and I'm 50:50 on re-using the variable name...

Are those your concerns, or are there other things? And given all of that, how do you feel about the variable naming thing?

Reply:
Ah okay that makes a lot of sense that you're destructuring. The confusing thing is reusing the name.

I strongly prefer:
```rust
let component_to_strip_opt = components.next();
if let Some(component_to_strip) = component_to_strip_opt {
```

Followed by

```rust
let maybe_component_to_strip = components.next();
if let Some(component_to_strip) = maybe_component_to_strip {
```

While Rust may allow the original with its move semantics, the rename makes the thing feel very mutation-oriented and the confusion is not worth saving the four extra characters of `_opt`.

Reply:
The rename (aka "shadowing") is a fairly rusty convention. The type system and borrow checker prevent (almost?) all of the errors that you can imagine from this pattern. And it definitely doesn't allow for mutation.

Reply:
Okay, I'll defer to you both here. I would rather we go with the Rusty convention.

Thanks for the thorough explanation! Learned something new.